### ISSUE TITLE
External trait interface doesn't compile

### ISSUE THREAD

**Issue by Voronar** [mozilla/uniffi-rs#1831]
I'm trying to reuse uniffi trait interface (https://mozilla.github.io/uniffi-rs/foreign_traits.html) from an external crate living in my workspace.

main_lib:
```rust
use std::sync::Arc;
use shared_lib::*;

pub fn use_common_trait(common_trait: Arc<dyn CommonTrait>) {
    common_trait.foo();
}

uniffi::include_scaffolding!("main_lib");
```

main_lib UDL:
```
namespace main_lib {
  void use_common_trait(CommonTrait common_trait);
};

[ExternalInterface="shared_lib"]
typedef extern CommonTrait;
```

shared_lib:
```rust
pub trait CommonTrait: Send + Sync {
    fn foo(&self);
}

uniffi::include_scaffolding!("shared_lib");
```

shared_lib UDL:
```
namespace shared_lib {};

[Trait]
interface CommonTrait {
    void foo();
};
```

Reproducible example: https://github.com/Voronar/uniffi_external_trait_interface
uniffi version: 0.25

**Comment by badboy** [comment]
Can you provide the actual error you're running into?

**Comment by Voronar** [comment]
I think the `dyn` somehow isn't generated during codegen phase in case of external trait interface type.

**Comment by Voronar** [comment]
Maybe the problem is here https://github.com/mozilla/uniffi-rs/blob/main/uniffi_bindgen/src/interface/ffi.rs#L110

**Comment by mhammond** [comment]
yeah - the problem here is that the trait is "external". I don't think an external type even *knows* if it is a trait or not - it probably needs to though.

**Comment by mhammond** [comment]
this should actually work ok using procmacros I think. There are a few options for UDL though, but all have some work.

**Comment by Voronar** [comment]
It compiles with proc-macros, but crashes at runtime for example with Kotlin script test:

```
thread '<unnamed>' panicked at /Users/user/.cargo/git/checkouts/uniffi-rs-6f89edd2a1ffa4bd/659d5d6/uniffi_core/src/ffi/foreigncallbacks.rs:102:1:
Bug: callback not set.  This is likely a uniffi bug.
```

Reproducible example: https://github.com/Voronar/uniffi_external_trait_interface/tree/external_trait_proc_macro_kotlin_script_crash

**Comment by mhammond** [comment]
The problem here is still UDL - main_lib's UDL can't reference the trait - so if you remove the trait and function from the UDL and use `#[uniffi::export]` on the function in lib.rs things seem to work. It would be possible to make UDL work here but it would probably require a contributor to implement it.

**Comment by mhammond** [comment]
See also #1854 and #1855

**Comment by Voronar** [comment]
> The problem here is still UDL...

I made reproducible example (mentioned in my previous message) without any UDL and it crashes at runtime.
> Reproducible example: https://github.com/Voronar/uniffi_external_trait_interface/tree/external_trait_proc_macro_kotlin_script_crash
